### PR TITLE
refs #22135 changed per-QOS logic to use pool pre-instantiated process

### PR DESCRIPTION
**Pull request by jkothamb**
In production, seeing growth in memory consumption of the main celery process. Putting memory profile didn't indicate any memory leak. Since the main celery process forks tenant poller that per-tenant QOS, the child process inherits the memory of the parent process which basically double the memory usage. The new approach taken is to use a python multi-processing pool with just a single process. The pool is instantiated on startup of the celery workers. The tenant poller task is given to this pool process for performing the per-tenant QOS logic.

### CODE REVIEW

**Comment by jkothamb** [comment]
@ChaksChigurupati : Can you please do the needful. Thanks.

**Comment by ChaksChigurupati** [comment]
@jkothamb : Can you resolve the merge conflicts?

**Comment by ChaksChigurupati** [comment]
@jkothamb : See the following exceptions in production logs. Looks like you need to check for existence before you call tenants.remove().

[2015-06-29 04:55:35,528: INFO/16548] Got task from broker: drive.drive_snaphot.SnapshotTask[939e3f5f-9e9e-4fac-8939-d3ade19e4234] eta:[2015-06-29 04:55:45.517289+00:00]
[2015-06-29 04:55:35,536: ERROR/13383] [tenant poller] Error occurred looping through the tailable cursor
Traceback (most recent call last):
  File "/home/madmin/venv/connectors/local/lib/python2.7/site-packages/kombu/transport/el_mongodb.py", line 195, in tenant_poller
    remove_tenants()
  File "/home/madmin/venv/connectors/local/lib/python2.7/site-packages/kombu/transport/el_mongodb.py", line 162, in remove_tenants
    tenants.remove(tenant)
KeyError: 'solucionesorioncom'
[2015-06-29 04:55:35,536: WARNING/13383] Process tenant_poller_drive_queue:
[2015-06-29 04:55:35,536: WARNING/13383] Traceback (most recent call last):
[2015-06-29 04:55:35,537: WARNING/13383] File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
[2015-06-29 04:55:35,537: WARNING/13383] self.run()
[2015-06-29 04:55:35,537: WARNING/13383] File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
[2015-06-29 04:55:35,537: WARNING/13383] self._target(*self._args, **self._kwargs)
[2015-06-29 04:55:35,537: WARNING/13383] File "/home/madmin/venv/connectors/local/lib/python2.7/site-packages/kombu/transport/el_mongodb.py", line 214, in tenant_poller
[2015-06-29 04:55:35,537: WARNING/13383] remove_tenants()
[2015-06-29 04:55:35,537: WARNING/13383] File "/home/madmin/venv/connectors/local/lib/python2.7/site-packages/kombu/transport/el_mongodb.py", line 162, in remove_tenants
[2015-06-29 04:55:35,538: WARNING/13383] tenants.remove(tenant)
[2015-06-29 04:55:35,538: WARNING/13383] KeyError: 'conexsysnet'

File: kombu/transport/el_mongodb.py [hunk 1]
```diff
@@ -79,8 +120,11 @@ def _get_task(self):
                 obj_id = str(latest_tenant_entry['_id'])
                 #logger.debug('Latest object id in tenant poller - %s' % str(latest_obj_id))
             
-            self.terminate_poller()
-            self.proc = None
+            #self.terminate_poller()
+            if self.async_result and not self.async_result.ready():
```
Comment (ChaksChigurupati) [inline kombu/transport/el_mongodb.py]: Why do you need to signal a perfectly running poller task to terminate? What guarantee is there that the poller will read this signal before the event is cleared at line 146?

File: kombu/transport/el_mongodb.py [hunk 2]
```diff
@@ -214,11 +248,8 @@ def remove_tenants():
                 logger.exception('[tenant poller] Error occurred in closing mongo connection')
         
         remove_tenants()        
-        # http://stackoverflow.com/questions/8594909/python-multiprocessing-pipe-recv-doc-unclear-or-did-i-miss-anything
-        child_pipe.close()
 
     logger.info('[tenant poller] exiting with pid %s' % str(os.getpid()))
```
Comment (ChaksChigurupati) [inline kombu/transport/el_mongodb.py]: Need to change this log? the process is never exiting but only the task is returning to caller now.

**Comment by jkothamb** [comment]
@ChaksChigurupati : Went with the idea of keeping the tenant poller task permanently running unless it exits due to some error which will result in new one to be scheduled. I could not instantiate a new event semaphore without having a new Pool instance also.

The KeyError was happening due the main process putting same tenant name multiple times in the pipe as it finds task for this tenant with a slight delay. I put in a try/catch block in the remove call

**Comment by jkothamb** [comment]
@ChaksChigurupati : Put try/catch for all remove operations

**Comment by ChaksChigurupati** [comment]
@jkothamb : Just for the record, I want to document our last discussion. The check on "len(self.tenants) == 0" can be replaced with a check on whether the poller process is already running. This will ensure that the main process hits Mongo only in the very beginning. After that it will be only the poller that will hit mongo on the tailable collection.

**Pull request merged**